feat(migrate-engine): start moving out of introspection-engine

jkomyno opened the pull request:
Closes https://github.com/prisma/schema-team/issues/375

**TODO**s after `prisma@4.9.0`:
- [x] https://github.com/prisma/prisma-private/issues/203

jkomyno committed:
feat(migrate-engine): start moving out of introspection-engine

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
feat(migration-engine): adapted compositeTypeDepth after https://github.com/prisma/prisma-engines/pull/3487 got merged

jkomyno committed:
feat(internals): stop using IntrospectionEngine.getDatabaseDescription, as it hasn't been used in years

jkomyno committed:
feat(integration-tests): replaced IntrospectionEngine with MigrateEngine

jkomyno committed:
feat(migration-engine): add "latestSchema" and "isRunning" properties

jkomyno committed:
feat(migration-engine): add "getDatabaseVersionSafe" util to avoid circular dependencies

jkomyno committed:
feat(cli): plugged "getDatabaseVersionSafe" into "handlePanic" and "sendPanic", avoiding circular dependencies in @prisma/internals. Standardized "engineVersion" into "enginesVersion"

jkomyno committed:
feat(integration-tests): depend on @prisma/migrate

jkomyno committed:
Revert "feat(migration-engine): adapted compositeTypeDepth after https://github.com/prisma/prisma-engines/pull/3487 got merged, as the Rust PR doesn't seem to have solved the type mistmatch issue"

This reverts commit d6052ec08214963f4d76fb108053c47bcc6d4810.

jkomyno commented on packages/internals/src/sendPanic.ts:
```diff
@@ -31,23 +43,11 @@ export async function sendPanic(error: RustPanic, cliVersion: string, engineVers
-          engine.stop()
-        }
-      }
+    if (error.area === ErrorArea.LIFT_CLI && schemaOrUrl) {
```
The `MigrateEngine` may throw Rust panics with the `LIFT_CLI` error area, rather than the previous `INTROSPECTION_CLI` error area thrown by the `IntrospectionEngine`

janpio replied:
No chance of this now triggered for ME errors as well? (probably undertested area)

janpio replied:
Ah, I see this is only to get the database version in a safe way - not more.

jkomyno commented on packages/internals/src/sendPanic.ts:
```diff
@@ -20,7 +19,20 @@ const debug = Debug('prisma:sendPanic')
 tmp.setGracefulCleanup()
 
-export async function sendPanic(error: RustPanic, cliVersion: string, engineVersion: string): Promise<number> {
+type SendPanic = {
+  error: RustPanic
+  cliVersion: string
+  enginesVersion: string
+
+  // retrieve the database version for the given schema or url, without throwing any error
+  getDatabaseVersionSafe: (schemaOrUrl: string) => Promise<string | undefined>
+}
+export async function sendPanic({
+  error,
+  cliVersion,
+  enginesVersion,
+  getDatabaseVersionSafe,
+}: SendPanic): Promise<number> {
```
Turning a list of arguments to an object with named properties for ease of use. It'd be easy passing the arguments in the wrong sequence instead.

jkomyno committed:
feat(introspection-engine): restore "introspect()" now that it's no longer used in the codebase, marking it as deprecated

jkomyno committed:
fix(migration-engine): corrected "EngineArgs.IntrospectParams" type

jkomyno commented:
Integration test errors should be fixed by https://github.com/prisma/prisma-engines/pull/3495.

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
chore(internals): update snapshot

janpio commented on packages/cli/src/Doctor.ts:
```diff
@@ -87,13 +86,14 @@ ${chalk.bold('Examples')}
-    const engine = new IntrospectionEngine({
-      cwd: path.dirname(schemaPath),
+    const engine = new MigrateEngine({
+      projectDir: path.dirname(schemaPath),
```
For later: Would `cwd` maybe be a better name for `projectDir` as well?

jkomyno replied:
That's a potential choice, yeah

janpio commented on packages/internals/src/sendPanic.ts:
```diff
@@ -20,7 +19,20 @@ const debug = Debug('prisma:sendPanic')
+type SendPanic = {
+  error: RustPanic
+  cliVersion: string
+  enginesVersion: string
```
I think the old name of `engineVersion` (singular, not plural) was chosen because it is the version of _one_ engine. In theory all engines in the CLI can use different versions, and this highlights that this is specifically the version from that one engine file.

jkomyno replied:
I'm glad we discussed this then! I'm creating a private issue about it

janpio commented on packages/internals/src/sendPanic.ts:
```diff
@@ -31,23 +43,11 @@ export async function sendPanic(error: RustPanic, cliVersion: string, engineVers
 
     const maskedSchema: string | undefined = schema ? maskSchema(schema) : undefined
 
-    let sqlDump: string | undefined
```
Not used before anyway so dead?

janpio replied:
Ah, I see below.

janpio commented on packages/internals/src/sendPanic.ts:
```diff
@@ -31,23 +43,11 @@ export async function sendPanic(error: RustPanic, cliVersion: string, engineVers
-      let engine: undefined | IntrospectionEngine
-      try {
-        engine = new IntrospectionEngine()
-        sqlDump = await engine.getDatabaseDescription(schemaOrUrl)
```
Are we sure that was not used for something and e.g. submitted with the error report?

jkomyno replied:
I had asked both Joel and Tom before removing `getDatabaseDescription` (which isn't implemented for MigrateEngine and isn't used anywhere else). The rationale is that we haven't made use of it in a long time (Tom never used it, Matthias did once or twice), and if we can remove legacy non-user-facing stuff, that's "always the best option"

janpio replied:
This referred to the resulting `sqlDump` - if we should not be using that to report to the error report. Might very well be that this system is actually set up to accept that and might benefit from doing so.

janpio commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -135,11 +142,29 @@ export class MigrateEngine {
-  public getDatabaseVersion(): Promise<string> {
-    return this.runCommand(this.getRPCPayload('getDatabaseVersion', undefined))
+  public getDatabaseVersion({ schema }: EngineArgs.GetDatabaseVersionParams): Promise<string> {
+    return this.runCommand(this.getRPCPayload('getDatabaseVersion', { schema }))
```
Why does this now also take/need the schema?

jkomyno replied:
The `getDatabaseVersion` RPC method had always required a schema. We didn't notice that the `getDatabaseVersion` recently added to MigrateEngine was missing an input argument because that method wasn't used in tests

janpio commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -135,11 +142,29 @@ export class MigrateEngine {
+  /**
+   * Given a Prisma schema, introspect the database definitions and update the schema with the results.
+   * `compositeTypeDepth` is optional, and only required for MongoDB.
+   * TODO: change `compositeTypeDepth` to -1 once https://github.com/prisma/prisma-engines/pull/3487 is merged.
```
PR https://github.com/prisma/prisma-engines/pull/3487 got dismissed

jkomyno replied:
I had done that and reverted it in https://github.com/prisma/prisma/pull/16766#:~:text=%E2%80%A6-,392a6d2,-jkomyno%20added%20this, as apparently that PR alone doesn't fix this type mismatch problem in JSON-RPC between JS and Rust

janpio commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -296,6 +324,8 @@ export class MigrateEngine {
           },
         })
 
+        this.isRunning = true
```
Now way that the code after this still makes it fail in some way that `isRunning=true` would be inaccurate?

jkomyno replied:
This `isRunning` properties is implemented mirroring the same property in the IntrospectionEngine, and is updated in the same way. The idea is to first get the MigrateEngine on par with the features exposed by the IntrospectionEngine we need to make use of, and later verify possible improvements that could be done

janpio commented on packages/migrate/src/utils/getDatabaseVersionSafe.ts:
```diff
@@ -0,0 +1,28 @@
+ * Retrieve the database version from the given schema or url.
+ * This function never throws, and was introduced to prevent circular dependencies in `@prisma/internals`.
+ */
+export async function getDatabaseVersionSafe(schemaOrUrl: string): Promise<string | undefined> {
```
When reading the code, this name threw me off, but I guess it makes sense to describe what it is doing.

janpio commented on packages/migrate/src/utils/getDatabaseVersionSafe.ts:
```diff
@@ -0,0 +1,28 @@
+  let engine: MigrateEngine | undefined
+  let dbVersion: string | undefined
+  try {
+    engine = new MigrateEngine({
+      projectDir: process.cwd(),
+    })
+    dbVersion = await engine.getDatabaseVersion({ schema: schemaOrUrl })
```
Does this always start up a new Engine to make this query, or would it reuse an existing one if it already exists?

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
fix(migrate): pass schema path to "getDatabaseVersion"

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
feat(migrate): understood "getDatabaseVersion"

jkomyno committed:
chore: removed comment

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
feat(migration-engine): temporarily removed support for "getDatabaseVersion" due to https://github.com/prisma/prisma-private/issues/203

Jolg42 committed:
test(integration-tests): close db connection after setting up sql

to avoid exhausting the max_connection with idle connections (on my computer = 100)

It was already done for SQL Server but not for PostgreSQL, MySQL, MariaDB

Jolg42 committed:
test(integration-tests): start and stop the engine for each test

The ME will keep connections until it's closed, which means that we could exhaust the max_con number of connection open for postgresql
see https://prisma-company.slack.com/archives/C02FNFLDUS3/p1671204843482509

Jolg42 committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
Merge branch 'feat/deprecate-introspection-engine-in-favor-of-migrate-engine' of github.com:prisma/prisma into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
chore: update snapshot

jkomyno committed:
chore: merge main, fix conflicts in pnpm-lock.yaml

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
Merge branch 'feat/deprecate-introspection-engine-in-favor-of-migrate-engine' of github.com:prisma/prisma into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

Jolg42 commented:
Note from Alberto: we need Tom to unblock the progress on this PR on engines side.

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
feat: shut down migration-engine once "introspect()" has returned

jkomyno committed:
Merge branch 'feat/deprecate-introspection-engine-in-favor-of-migrate-engine' of github.com:prisma/prisma into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

jkomyno committed:
fix: "should fail when db is empty" sqlite test, changed P1003 in favor of P4001

jkomyno committed:
chore: add comments about sqlite

Jolg42 committed:
fix engine.stop()

jkomyno committed:
feat: kill the migration-engine after a successful or failing introspection

jkomyno committed:
feat: fix artificial-panic test

jkomyno committed:
feat: fix incomplete-schemas test

Jolg42 commented on packages/migrate/src/commands/DbPull.ts:
```diff
@@ -252,9 +252,9 @@ Some information will be lost (relations, comments, mapped fields, @ignore...),
     } catch (e: any) {
       introspectionSpinner.failure()
-      if (e.code === 'P4001') {
+      if (e.code === 'P1003') {
         if (introspectionSchema.trim() === '') {
-          throw new Error(`\n${chalk.red.bold('P4001 ')}${chalk.red('The introspected database was empty:')} ${
+          throw new Error(`\n${chalk.red.bold('P1003 ')}${chalk.red('The introspected database was empty:')} ${
```
I think that this shows a breaking change in our public API here before we had the error code `P4001` and now it disappears (changed to P1003)
We need to find another way and keep the old codes or go for a major version bump.

jkomyno replied:
Answered here: https://prisma-company.slack.com/archives/C04HZ2EQS65/p1673446997564369?thread_ts=1673440456.031899&cid=C04HZ2EQS65

Jolg42 replied:
TODO:
We now we have 2 cases
- Database is empty
- Database does not exist
We need to check if we have tests for both in introspection

jkomyno replied:
I've added human-friendly error support for both `P4001` and `P1003`. Please check https://github.com/prisma/prisma/pull/16766/commits/9c820ae3348de82991defb068750189426bbfaa5

jkomyno committed:
feat(migrate): add support for both P1003 and P4001, add "postgresql - missing database" testcd '/Users/jkomyno/work/prisma/prisma/packages/migrate'

jkomyno committed:
chore: merge main, fix conflicts

prisma-bot committed:
chore(deps): update engines to 4.9.0-33.bb177fac1c08b8561c02b04d9ad51a31981cd8a3

Jolg42 commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -778,7 +814,7 @@ describe('postgresql-multi-schema', () => {
     expect(ctx.mocked['process.stdout.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
     expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
-  })
+  }, 50_000)
```
Why is this skipped now?

jkomyno replied:
Heck, this was unintentional, I had CMD+F'd the "skip" string in https://github.com/prisma/prisma/pull/16766/files without realizing `DbPull.test.ts` was compressed due to the diff being too large

jkomyno replied:
Thankfully the test passes anyway

Jolg42 commented on pnpm-lock.yaml:
```diff
@@ -229,7 +229,7 @@ importers:
       '@prisma/engine-core': workspace:*
       '@prisma/engines': workspace:*
-      '@prisma/engines-version': 4.9.0-32.f307669e852dde0c751a36420b11f932fc3cf268
+      '@prisma/engines-version': 4.9.0-33.bb177fac1c08b8561c02b04d9ad51a31981cd8a3
```
Diff in the pnpm lock is unexpected, did you install this engine before it landed in main?

jkomyno replied:
Yes, I had tried it out before it landed in main to see whether all tests passed. I can revert https://github.com/prisma/prisma/pull/16766/commits/fcbd3bcf9cf70db9125b12ea70b59739d2fc0243 and merge `main` back to this branch

Jolg42 replied:
It's probably fine in the end but you will get a conflict in this PR once a new engine version lands

jkomyno replied:
I have already done the steps described in my previous message with no conflicts whatsoever :)

jkomyno commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -412,6 +414,8 @@ describe('common/sqlite', () => {
+  // The error tells us to try --force, which we do already. Perhaps --force isn't passed properly to Rust?
   it('should succeed when schema is invalid and using --force', async () => {
     ctx.fixture('introspect')
 
```
As of yesterday evening, not anymore. We can get rid of it

Jolg42 commented on packages/migrate/src/utils/test-handlePanic.ts:
```diff
@@ -11,9 +11,16 @@ async function main() {
   const packageJsonVersion = '0.0.0'
-  const engineVersion = '734ab53bd8e2cadf18b8b71cb53bf2d2bed46517'
+  const enginesVersion = '734ab53bd8e2cadf18b8b71cb53bf2d2bed46517'
+  const getDatabaseVersionSafe = () => Promise.resolve(undefined)
```
Could you use the one from `getDatabaseVersionSafe.ts` ?

jkomyno replied:
I see that the same test in `packages/cli` uses the literal `prismaEngineVersionHash` (https://github.com/prisma/prisma/blob/dba8c89d305d91b7cdd6cac6fea49f5192d39af3/packages/cli/src/utils/test-handlePanic.ts#L8), I may just use the same in `packages/migrate`

Jolg42 replied:
Actually this file is not even used anywhere, it's a small script that someone can run locally (not part of tests)

So anything works I guess, but let's still remove `const getDatabaseVersionSafe = () => Promise.resolve(undefined)` so we don't forget later

jkomyno replied:
Why would we remove `const getDatabaseVersionSafe = () => Promise.resolve(undefined)`? It's purposedly a mocked function

jkomyno committed:
Revert "chore(deps): update engines to 4.9.0-33.bb177fac1c08b8561c02b04d9ad51a31981cd8a3"

This reverts commit fcbd3bcf9cf70db9125b12ea70b59739d2fc0243.

jkomyno committed:
Merge branch 'main' into feat/deprecate-introspection-engine-in-favor-of-migrate-engine

Jolg42 commented on packages/migrate/src/types.ts:
```diff
@@ -137,6 +137,197 @@ export namespace EngineArgs {
+    code: 19
+    affected: AffectedModel[]
+  }
+
```
It's missing
```
interface IntrospectionWarningsTopLevelItemNameIsADupe extends IntrospectionWarning {
  code: 20
  affected: AffectedTopLevel[]
}
```

Can you add the recent changes from this commit?
https://github.com/prisma/prisma/commit/c2d32eea77e963778ba0b98cc3557ad0de62a5b7#diff-a3ca29a928d7e7c324513f5239c93d68b07cbd190c678587efc3eb2a93c32d82

Jolg42 replied:
Note: It's the only change since this PR was opened

jkomyno replied:
Done

jkomyno committed:
chore: undo test.skip

jkomyno committed:
chore: nit on enginesVersion

jkomyno committed:
chore: removed outdated comment

jkomyno committed:
feat: add missing types from c2d32eea77e963778ba0b98cc3557ad0de62a5b7

Jolg42 commented on packages/cli/schema.prisma:
```diff
@@ -0,0 +1,14 @@
+datasource db {
```
This schema file at the root of the CLI package is unexpected to me.
Could you remove it? (Is it needed for anything?)

Jolg42 commented on packages/migrate/src/__tests__/introspection/introspection.test.ts:
```diff
@@ -0,0 +1,55 @@
+  // const dbVersion = await engine.getDatabaseVersion({ schema })
+  // expect(dbVersion.length > 0).toBe(true)
+
+  engine.stop()
```
Still needed?

janpio replied:
Is this still needed?

Jolg42 commented on packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:
```diff
@@ -310,7 +307,12 @@ async function setupScenario(kind: string, input: Input, scenario: Scenario) {
+    projectDir: process.cwd(),
+  })
+  const introspectionResult = await engine.introspect({ schema: schemaBase })
+  engine.stop()
```
Still needed?

Jolg42 commented on packages/internals/src/__tests__/introspection/introspection.test.ts:
```diff
@@ -70,6 +64,5 @@ test('introspection basic', async () => {
   const json = JSON.parse(description)
   expect(typeof json).toBe('object')
-
   engine.stop()
```
Still needed?

Jolg42 replied:
I think it is actually needed here

janpio commented on packages/cli/src/__tests__/artificial-panic.test.ts:
```diff
@@ -48,8 +56,7 @@ describe('artificial-panic introspection', () => {
         expect(e).toMatchObject({
-          area: 'INTROSPECTION_CLI',
-          schemaPath: undefined,
+          area: 'LIFT_CLI',
```
If now introspection errors also report `LIFT_CLI`, do we still have enough information in the error reporting tool to quickly place where the error comes from?

jkomyno replied:
Yes, once https://github.com/prisma/prisma-private/issues/203 is closed (`prisma@4.10.0`)

janpio replied:
How does that influence it?

jkomyno replied:
Besides the actual database version (which is WIP) and the error area (which doesn't matter), this PR will preserve every panic information as before

janpio replied:
> and the error area (which doesn't matter)

There is literally a filter by error area in the error report tool.

Jolg42 commented on packages/internals/src/IntrospectionEngine.ts:
```diff
@@ -269,6 +269,10 @@ export class IntrospectionEngine {
   }
+
+  /**
+   * @deprecated Use `MigrateEngine.introspect()` instead
```
```suggestion
   * @deprecated Use `MigrateEngine.introspect()` instead from `@prisma/migrate` package
```

janpio commented on packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:
```diff
@@ -310,7 +307,12 @@ async function setupScenario(kind: string, input: Input, scenario: Scenario) {
+    projectDir: process.cwd(),
+  })
+  const introspectionResult = await engine.introspect({ schema: schemaBase })
+  engine.stop()
```
Is this really necessary here? Doesn't `introspect()` take care of that itself?

jkomyno replied:
I think @Jolg42 ninjed you, I have fixed this already while you were writing this :)

janpio replied:
I can live with that.

janpio commented on packages/integration-tests/src/__tests__/integration/mariadb/__database.ts:
```diff
@@ -25,6 +25,7 @@ export const database = {
     CREATE DATABASE ${ctx.id};
     USE ${ctx.id};`
     await db.query(sqlUp + sqlScenario)
+    await db.end()
```
Why is this needed?

jkomyno replied:
I'll be honest, in retrospect it doesn't seem useful

jkomyno replied:
`git blame` tells me that @Jolg42 added it

Jolg42 replied:
Yes it's a nice improvement we already did that for SQL Server only I think, found out when we had the exhaustion of connections on PostrgeSQL

janpio commented on packages/integration-tests/src/__tests__/integration/mysql/__database.ts:
```diff
@@ -24,6 +24,7 @@ export const database = {
     CREATE DATABASE ${ctx.id};
     USE ${ctx.id};`
     await db.query(sqlUp + sqlScenario)
+    await db.end()
```
Why is this needed?

janpio commented on packages/integration-tests/src/__tests__/integration/postgresql/__database.ts:
```diff
@@ -19,6 +19,7 @@ export const database = {
     create schema ${ctx.id};
     set search_path to ${ctx.id};`
     await db.query(sqlUp + sqlScenario)
+    await db.end()
```
Why is this needed?

janpio commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -43,6 +43,13 @@ export class MigrateEngine {
   private enabledPreviewFeatures?: string[]
+
+  // `latestSchema` is set to the latest schema that was used in `introspect()`
+  private latestSchema?: string
```
Is this only used for Introspection? Then it might make sense to include that in the name maybe. `latestIntrospectedSchema` for example.

jkomyno replied:
Done

janpio commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -135,11 +142,42 @@ export class MigrateEngine {
+
   /**
    * Get the database version for error reporting.
+   * - TODO: this command currently requires the engine to be initialized with a schema path.
+   *   The IntrospectionEngine supported reading a database version from an inline schema or URL instead,
+   *   which if more flexible and fundamental for the error reporting use case. We should add that here too.
```
Can we already create an issue for this?

jkomyno replied:
There's already an issue linked in the comments

janpio replied:
That seemed to be a separate TODO, so I expected a different issue. 
(The other issue is also not really about that TODO, but a requesite for that TODO)

jkomyno committed:
chore: removed useless file

jkomyno committed:
chore: removed redundant engine.stop()

jkomyno committed:
chore: removed redundant engine.stop()

janpio commented on packages/migrate/src/MigrateEngine.ts:
```diff
@@ -458,6 +503,8 @@ export class MigrateEngine {
 
 /** The full message with context we return to the user in case of engine panic. */
 function serializePanic(log: string): string {
+  // TODO: do we still want the "Please create an issue ..." part? That's supposedly
+  // already handled in cli/bin.ts when a panic occurs.
```
Issue?

jkomyno replied:
```suggestion
  // TODO: https://github.com/prisma/prisma/issues/17268
```

jkomyno committed:
chore: recovered engine.stop() in introspection.test.ts

Jolg42 commented on packages/internals/src/__tests__/handlePanic.test.ts:
```diff
@@ -119,14 +134,20 @@ describe('handlePanic', () => {
       'test-message',
       rustStackTrace,
       'test-request',
-      ErrorArea.INTROSPECTION_CLI, // area
+      ErrorArea.LIFT_CLI, // area
```
Not sure what we want to do here, it's still about introspection but the code leaves in the migration engine.

jkomyno replied:
The `RustPanic` in `MigrateEngine` are statically defined as `LIFT_CLI`. Although it's technically introspection-related, this tells us the engine that was used under the hood. Besides, this `ErrorArea` label is of internal-use only

Jolg42 replied:
`INTROSPECTION_CLI` would be better to keep for the meaning of the error was actually triggered from the "introspection part of the CLI"
But since it would be complicated and would not provide any benefit, we can go for `LIFT_CLI`
Technically, we could also remove completely the `ErrorArea` one day (but we need to handle that in the error reporting backend)

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -169,9 +169,11 @@ describe('common/sqlite', () => {
     const introspect = new DbPull()
     const result = introspect.parse(['--print', '--url', 'postgresql://root:prisma@/prisma'])
     await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
-            Error parsing connection string: empty host in database URL
```
Nice improvement that this is a proper error now!

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -169,9 +169,11 @@ describe('common/sqlite', () => {
+      P1013
 
-          `)
+      The provided database string is invalid. empty host in database URL. Please refer to the documentation in https://www.prisma.io/docs/reference/database-reference/connection-urls for constructing a correct connection string. In some cases, certain characters must be escaped. Please check the string for any illegal characters.
```
This message is not optimal, but I guess unrelated to this PR and also not super important.

jkomyno replied:
Yep, this is unrelated to this PR

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -427,19 +429,51 @@ describe('common/sqlite', () => {
 
-                                                            ✔ Introspected 3 models and wrote them into prisma/invalid.prisma in XXXms
-                                                                  
-                                                            Run prisma generate to generate Prisma Client.
+                                                                        ✔ Introspected 3 models and wrote them into prisma/invalid.prisma in XXXms
+                                                                              
+                                                                        Run prisma generate to generate Prisma Client.
 
-                                        `)
+                                                `)
```
OT: One day I hope we will figure out what adds these useless spaces all the time and can fix it.

jkomyno replied:
I agree. Using jest with the `-u` flag causes this, from time to time. Not sure why exactly it happens, though

jkomyno committed:
[skip ci] chore: nit comment

Co-authored-by: Joël Galeran <Jolg42@users.noreply.github.com>

Jolg42 commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -427,19 +429,51 @@ describe('common/sqlite', () => {
+    const result = introspect.parse(['--print', '--url', connectionString])
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+      P1003 The introspected database does not exist: postgres://prisma:prisma@localhost:5432/unknown-database
```
Unrelated to this PR I guess: I'm not a fan of printing the user and password to the logs 😨 
I'll create an issue for this

Jolg42 replied:
It's only for --url, still undocumented so it's not a blocker

jkomyno replied:
I think we already handle secretation elsewhere. Perhaps this should be a function defined just once in the Rust engines and exported via Wasm. Can you please open an issue for this?

Jolg42 replied:
https://github.com/prisma/prisma/issues/17286

jkomyno commented on packages/migrate/src/__tests__/DbPush.test.ts:
```diff
@@ -190,7 +190,7 @@ describe('push', () => {
     `)
     expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
-    expect(mockExit).toBeCalledWith(130)
+    expect(mockExit).toHaveBeenCalledWith(130)
```
They are behaviorally the same, but the new one was suggested by the Jest extension I'm using in VSCode, which leads me to believe it's the most common "expect" method for that task. It's also compatible with the standard `jasmine` nomenclature, adopted by all other test frameworks. I didn't really notice this change as it was an automated one

jkomyno commented on packages/migrate/src/__tests__/MigrateDiff.test.ts:
```diff
@@ -30,35 +30,35 @@ describe('migrate diff', () => {
 
       const result = MigrateDiff.new().parse([])
-      await expect(result).rejects.toThrowError()
+      await expect(result).rejects.toThrow()
```
Same as above

Jolg42 reviewed (approved):
Nice! Let's do this 🚀

jkomyno committed:
chore: add issue

janpio commented:
Let's create a public `prisma/prisma` issue for this that will show up in our release notes so people can learn about this.

Jolg42 commented:
Note: first appeared in `4.9.0-dev.55`
